File: 백준/Platinum/12963. 달리기/달리기.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] parent;
    static int MOD = 1_000_000_007;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        ArrayList<long[]> roads = new ArrayList<>();
        long result = 0;

        // parent 초기화
        parent = new int[N];
        for (int i = 0; i < N; i++) {
            parent[i] = i;
        }

        long multiply = 1;
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            roads.add(new long[] {a, b, multiply});

            multiply *= 3;
            multiply %= MOD;
        }

        for (int i = M-1; i >= 0; i--) {
            long[] tmp = roads.get(i);
            long a = tmp[0]; long b = tmp[1]; long cost = tmp[2];

            int pa = find((int) a); int pb = find((int) b);
            int ps = find(0); int pe = find(N-1);

            if ((pa == ps && pb == pe) || (pa == pe && pb == ps)) {
                result += cost;
                result %= MOD;
            } else {
                parent[pa] = pb;
            }
        }
        System.out.println(result);
    }

    static int find(int n) {
        if (parent[n] != n) {
            parent[n] = find(parent[n]);
        }

        return parent[n];
    }
}

File: 백준/Platinum/12963. 달리기/달리기.py
# 달리기
import sys
input = sys.stdin.readline

def find(n):
    if parent[n] != n:
        parent[n] = find(parent[n])

    return parent[n]

N, M = map(int,input().split())
roads = []
parent = [i for i in range(N)]
result = 0
MOD = 1_000_000_007

multiply = 1
for i in range(M):
    a, b = map(int,input().split())
    roads.append((a, b, multiply))
    multiply *= 3

for i in range(M-1, -1, -1):
    a, b, cost = roads[i]

    pa, pb = find(a), find(b)
    ps, pe = find(0), find(N-1)

    if ((pa == ps and pb == pe) or (pa == pe and pb == ps)):
        result += cost
        result %= MOD
    else:
        parent[pa] = pb

print(result)